plugins/logitech-tap/fu-logitech-tap-hdmi-device.c

```diff
@@ -0,0 +1,522 @@
+static gboolean
+fu_logitech_tap_hdmi_device_ait_finalize_update(FuLogitechTapHdmiDevice *self, GError **error)
+{
+	guint duration_ms = 0;
+	guint8 ait_finalize_update[XU_INPUT_DATA_LEN] =
+	    {kLogiTapAitSetMmpCmdFwBurning, kLogiTapVideoAitInitiateSetMMPData, 0, 0, 0, 0, 0, 0};
+
+	g_debug("Ait finalize update request");
+
+	fu_device_sleep(FU_DEVICE(self), 4 * kLogiDefaultAitSleepIntervalMs); /* 4 sec */
+	if (!fu_logitech_tap_hdmi_device_set_xu_control(self,
+							kLogiUnitIdVidCapExtension,
+							kLogiTapUvcXuAitCustomCsSetMmp,
+							XU_INPUT_DATA_LEN,
+							(guchar *)&ait_finalize_update,
+							error))
+		return FALSE;
+
+	fu_device_sleep(FU_DEVICE(self), kLogiDefaultAitSleepIntervalMs); /* 1 sec */
+	/* poll for burning fw result or return failure if it hits max polling */
+	for (int pass = 0;; pass++) {
+		g_autofree guint8 *mmp_get_data = NULL;
+		guint16 data_len = 0;
+
+		fu_device_sleep(FU_DEVICE(self), kLogiDefaultAitSleepIntervalMs); /* 1 sec */
+		duration_ms = duration_ms + kLogiDefaultAitSleepIntervalMs;
+		if (!fu_logitech_tap_hdmi_device_query_data_size(
+			self,
+			kLogiUnitIdVidCapExtension,
+			kLogiTapUvcXuAitCustomCsGetMmpResult,
+			&data_len,
+			error))
+			return FALSE;
+		mmp_get_data = g_malloc0(data_len);
+		if (!fu_logitech_tap_hdmi_device_get_xu_control(
+			self,
+			kLogiUnitIdVidCapExtension,
+			kLogiTapUvcXuAitCustomCsGetMmpResult,
+			data_len,
+			(guchar *)mmp_get_data,
+			error))
+			return FALSE;
+		if (mmp_get_data[0] == kLogiDefaultAitSuccessValue) {
+			if (pass == 0)
+				g_usleep(8 * G_USEC_PER_SEC);
+			break;
+		} else if (mmp_get_data[0] == kLogiDefaultAitFailureValue) {
+			g_set_error(error,
+				    G_IO_ERROR,
+				    G_IO_ERROR_FAILED,
+				    "failed to finalize image burning, invalid result data: 0x%x",
+				    (guchar)mmp_get_data[0]);
+			return FALSE;
+		}
+		if (duration_ms > kLogiDefaultAitFinalizeMaxPollingDurationMs) {
+			/* if device never returns 0x82 or 0x00, bail out */
+			g_set_error(error,
+				    G_IO_ERROR,
+				    G_IO_ERROR_FAILED,
+				    "failed to finalize image burning, duration_ms: %u",
+				    duration_ms);
+			return FALSE;
+		}
+	}
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_logitech_tap_hdmi_device_write_fw(FuDevice *device,
+				     GPtrArray *chunks,
+				     FuProgress *progress,
+				     GError **error)
+{
+	FuLogitechTapHdmiDevice *self = FU_LOGITECH_TAP_HDMI_DEVICE(device);
+
+	/* init */
+	if (!fu_logitech_tap_hdmi_device_ait_initiate_update(self, error))
+		return FALSE;
+
+	/* write */
+	fu_progress_set_id(progress, G_STRLOC);
+	fu_progress_set_steps(progress, chunks->len);
+	for (guint i = 0; i < chunks->len; i++) {
+		FuChunk *chk = g_ptr_array_index(chunks, i);
+		g_autoptr(GByteArray) data_pkt = g_byte_array_new();
+		/* device expect fixed size buffer, so cannot leverage actual size here:
+		 * fu_chunk_get_data_sz(chk) */
+		g_byte_array_append(data_pkt, fu_chunk_get_data(chk), kLogiDefaultImageBlockSize);
```

Comment:
is fu_chunk_get_data_sz(chk) always bigger than kLogiDefaultImageBlockSize? do we need to allocate a [larger] buffer and copy in the chunk data perhaps?

Reply:
Not always, last packet going to be smaller. Device always expect 32 size chunks, expect padding if needed (e.g. last packet)

Reply:
ohh you definitely need to allocate and copy then -- otherwise we're sending random memory to the device.

Reply:
Fixed